### PR TITLE
Feature / block history

### PR DESCRIPTION
**Pull request by axelsheva**

### CODE REVIEW

File: core/repository/history/index.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import config from 'shared/config';
+
+export class HistoryRepository<T> {
+    store: Map<string, Array<T>>;
+
+    constructor() {
+        this.store = new Map();
+    }
+
+    add = (id: string, event: T) => {
```
Comment (Verbalist) [inline core/repository/history/index.ts]: in one style use `add(id: string, event: T)` for methods

File: core/repository/history/index.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        if (!config.CORE.IS_HISTORY) {
+            return;
+        }
+
+        const history = this.store.get(id);
+
+        history
+            ? this.store.get(id).push(event)
```
Comment (Verbalist) [inline core/repository/history/index.ts]: this.store.get(id) -> history

File: core/repository/history/index.ts (hunk 1 above)
Comment (Verbalist) [reply core/repository/history/index.ts]: `add` must return void

File: core/repository/history/transaction.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+import { HistoryRepository } from 'core/repository/history';
+
+class TransactionHistoryRepository extends HistoryRepository<TransactionHistoryEvent> {
+    addBeforeState(id: TransactionId, action: TransactionLifecycle, account: Account): void {
+        if (!config.CORE.IS_HISTORY) {
+            return;
+        }
+
+        if (!this.get(id)) {
```
Comment (Verbalist) [inline core/repository/history/transaction.ts]: ?

File: core/repository/history/transaction.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            this.add(id, { action, accountStateBefore: account.getCopy() });
+        }
+    }
+
+    addAfterState(id: TransactionId, action: TransactionLifecycle, account: Account): void {
+        if (!config.CORE.IS_HISTORY) {
+            return;
+        }
+
+        const history = this.get(id) || this.add(id, { action });
```
Comment (Verbalist) [inline core/repository/history/transaction.ts]: enough `const history = this.get(id)`

File: core/repository/history/index.ts (hunk 1 above)
Comment (Verbalist) [reply core/repository/history/index.ts]: event: T -> { entity: E, events: T[] }

File: core/repository/history/index.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    constructor() {
+        this.store = new Map();
+    }
+
+    addEvent(entity: Entity, event: EventType): void {
+        if (!config.CORE.IS_HISTORY) {
+            return;
+        }
+
+        console.log(`entity.id`, entity.id, `event`, event);
```
Comment (macaqueee) [inline core/repository/history/index.ts]: remove `console.log`

File: core/service/block.ts [hunk 6]
```diff
@@ -553,13 +545,14 @@ class BlockService {
...
         const blockTransactions = transactions.sort(transactionSortFunc);
-        return new Block({
+        const block = new Block({
             createdAt: timestamp,
             transactionCount: blockTransactions.length,
             previousBlockId: previousBlock.id,
             generatorPublicKey: keyPair.publicKey.toString('hex'),
             transactions: blockTransactions
         });
+        return block;
```
Comment (macaqueee) [inline core/service/block.ts]: why this redundant variable?

File: core/service/block.ts [hunk 7]
```diff
@@ -454,22 +453,14 @@ class BlockService {
         block.signature = this.calculateSignature(block, keyPair);
         block.id = this.getId(block);
 
-        return new ResponseEntity<Block>({ data: block });
-    }
+        BlockHistoryRepository.addEvent(block, { action: BlockLifecycle.CREATE });
```
Comment (macaqueee) [inline core/service/block.ts]: do we need object with field `action`?

**Pull request merged**